compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1532,11 +1532,85 @@ private boolean checkSimilarMatchGuard(BLangMatchGuard firstMatchGuard, BLangMat
             BLangTypeTestExpr secondTypeTest = (BLangTypeTestExpr) secondMatchGuard.expr;
             return ((BLangSimpleVarRef) firstTypeTest.expr).variableName.toString().equals(
                     ((BLangSimpleVarRef) secondTypeTest.expr).variableName.toString()) &&
-                    firstTypeTest.typeNode.getBType().tag == secondTypeTest.typeNode.getBType().tag;
+                    checkSimilarTypeDescInTypeTest(firstTypeTest.typeNode.getBType(),
+                            secondTypeTest.typeNode.getBType());
         }
         return false;
     }
 
+    private boolean checkSimilarTypeDescInTypeTest(BType firstType, BType secondType) {
+        if (firstType.tag != secondType.tag) {
+            return false;
+        }
+        switch (firstType.tag) {
+            case TypeTags.FINITE:
+                BFiniteType firstFiniteType = (BFiniteType) firstType;
+                BFiniteType secondFiniteType = (BFiniteType) secondType;
+                if (firstFiniteType.getValueSpace().size() != secondFiniteType.getValueSpace().size()) {
```

Comment:
Since we don't remove duplicates/flatten types, this may not always produce correct results, right?

e.g.,
```ballerina
type Foo 1|2;
type Bar 1|1|2;
```

Reply:
Same applies for unions too.

Can't we just check assignability for all instead of types.isSameType at the end?

Reply:
Fixed in ec5007c